lib/Importer/Caffe2ModelLoader.cpp

```diff
@@ -527,6 +527,33 @@ llvm::Error Caffe2ModelLoader::loadConvQuantized(const caffe2::OperatorDef &op,
   return llvm::Error::success();
 }
 
+llvm::Expected<bool>
+Caffe2ModelLoader::foldOperator(const caffe2::OperatorDef &op) {
+  const unsigned numInputs = op.input_size();
+  const std::string &typeName = op.type();
+  llvm::SmallVector<NodeValue, 4> inputs;
+  inputs.reserve(numInputs);
+  for (unsigned i = 0; i < numInputs; i++) {
+    NodeValue in;
+    ASSIGN_VALUE_OR_RETURN_ERR(in, getNodeValueByName(op.input(i)));
+    inputs.push_back(in);
+  }
+
+  if (!isConstantFoldable(inputs, typeName)) {
+    return false;
+  }
+
+  // Create a temporary lightweight loader to construct function representing
+  // current Op, and then constant fold the function using Interp backend.
+  Function *pfuncCF = G_.getParent()->createFunction("eval_const_fold__");
+  Caffe2ModelLoader loaderCF(*pfuncCF, nullptr);
+  bool fold_status =
+      !errToBool(constantFoldInLoader<Caffe2ModelLoader, caffe2::OperatorDef>(
```

Comment:
It's a little funny that we convert this err to a bool even though we're inside a function that returns an `llvm::Expected`. I see we have `ASSIGN_VALUE_OR_RETURN_ERR()` above. I think it would make sense to either return the error here, or otherwise above return false if `getNodeValueByName()` returns an error above and change this function to just return a `bool`.

Reply:
Problem with ASSIGN_VALUE_OR_RETURN_ERR is that it doesn't give the opportunity to do cleanup. We have to erase the temporary function which we created just before constantFoldInLoader. Hence the use of errToBool. ASSIGN_VALUE_OR_RETURN_ERR used above doesn't require cleanup hence is fine.